/****************************************************************************
 *
 * Cass80 tool - Main window class
 *
 * Copyright (C) 2020 Jürgen Buchmüller <pullmoll@t-online.de>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ****************************************************************************/
#pragma once
#include <QMainWindow>

QT_BEGIN_NAMESPACE
namespace Ui { class Cass80Main; }
QT_END_NAMESPACE

class Cass80Handler;
class bdfData;

class Cass80Main : public QMainWindow
{
    Q_OBJECT

public:
    Cass80Main(QWidget *parent = nullptr);
    ~Cass80Main();

private slots:
    void Info(QString message);
    void Error(QString message);

    bool about();
    bool load();
    bool save();
    bool information();
    bool undo_lmoffset();
    bool preferences();

private:
    void setup_toolbar();
    void setup_preferences();
    void update_actions();

    void set_listing(const QStringList& list);
    void set_listing(const QString& str);
    void update_cursor();
    void reset();
    Ui::Cass80Main *ui;
    Cass80Handler *m_cas;
    bdfData *m_bdf;
    bool m_internal_ttf;
    bool m_uppercase;
    QString m_filepath;
};
